usersession/userd/launcher.go

```diff
@@ -123,6 +130,157 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+	}
+
+	// Invoke indirectly via sh to unparent and avoid potentially leaving a zombie
+	cmd := exec.Command("sh", "-c", strings.Join(args, " ")+"&")
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		if !strutil.ListContains(allowedEnvVars, strings.SplitN(e, "=", 2)[0]) {
+			return dbus.MakeFailedError(fmt.Errorf("Supplied environment variable %q is not allowed", e))
+		}
+
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	if cmd.Run() != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+type fileExists func(string) bool
+
+func existsOnFileSystem(desktop_file string) bool {
+	fileStat, err := os.Stat(desktop_file)
+
+	return err == nil && !fileStat.IsDir()
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+func findDesktopFile(desktop_file_exists fileExists, base_dir string, splitFileId []string) *string {
+	desktop_file := filepath.Join(base_dir, strings.Join(splitFileId, "-"))
+
+	if desktop_file_exists(desktop_file) {
+		return &desktop_file
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID
+	for i := 1; i != len(splitFileId); i++ {
+		desktop_file := findDesktopFile(desktop_file_exists, filepath.Join(base_dir, strings.Join(splitFileId[:i], "-")), splitFileId[i:])
+		if desktop_file != nil {
+			return desktop_file
+		}
+	}
+
+	return nil
+}
+
+// desktopFileIdToFilename determines the path associated with a desktop file ID.
+func desktopFileIdToFilename(desktop_file_exists fileExists, desktop_file_id string) (string, error) {
+
+	// Currently the caller only has access to /var/lib/snapd/desktop/applications/, so we just look there
+	// and ignore https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	base_dir := dirs.SnapDesktopFilesDir
+
+	desktop_file := findDesktopFile(desktop_file_exists, base_dir, strings.Split(desktop_file_id, "-"))
+
+	if desktop_file != nil {
+		return *desktop_file, nil
+	}
+
+	return "", fmt.Errorf("cannot find desktop file for %q", desktop_file_id)
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry.
+func readExecCommandFromDesktopFile(desktop_file string) (string, error) {
+	var launch string
+
+	file, err := os.Open(desktop_file)
+	if err != nil {
+		return launch, err
+	}
+	defer file.Close()
+	reader := bufio.NewReader(file)
+
+	in_desktop_section := false
+	for {
+		line, err := reader.ReadString('\n')
+		if err != nil {
+			return launch, err
+		}
+
+		line = strings.TrimSpace(line)
+
+		if line == "[Desktop Entry]" {
+			in_desktop_section = true
+		} else if strings.HasPrefix(line, "[Desktop Action") {
+			// maybe later we'll add support here
+			in_desktop_section = false
+		} else if strings.HasPrefix(line, "[") {
+			in_desktop_section = false
+		} else if in_desktop_section && strings.HasPrefix(line, "Exec=") {
+			launch = strings.TrimPrefix(line, "Exec=")
```

Comment:
There's also some code in `usersession/autostart/autostart.go` that goes through the desktop file to find the command. It's bit more simplistic and just looks for the first entry that starts with `Exec=` prefix. Perhaps we could do just that instead?

Reply:
AFAIK there's no guarantee that "Desktop Entry" precedes "Desktop Action ...".

I know the Multipass team, and likely others, hope to extend support to include Desktop Actions. Following the desktop entry specification seems prudent.